### ISSUE TITLE
Updating values in lang/contexts.json make translations:upload purge existing translations

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-dev#377]
Change in context.json: https://github.com/ckeditor/ckeditor5-heading/commit/e4107c01c9ebef611e09adf4c06194f87d7132c1

I executed `translations:collect` and `translations:upload` and this is what I saw:

```
-------------------------------
Package: ckeditor5-heading
New: 6
Updated: 6
Deleted: 6
-------------------------------
```

And the entire ckeditor5-heading resource on Tx was empty. Tx completely lost track of the old strings and translations. Or we sent something incorrectly.

**Comment by ma2ciek** [comment]
To be precise, Tx lost track to all translations which their `msgctxt` changed, so, actually to all of them 😄

To revert that change we should:
1. run `npm run translations:collect`
2. manually change original PO file(s) in `ckeditor5/build/.transifex/`
3. run `npm run translations:upload`

I think, that the second script from https://github.com/ckeditor/ckeditor5-dev/issues/376 should protect us from similar issues in the future. It would allow us to update all translations, not only update English strings.

**Comment by ma2ciek** [comment]
And I've found that one: https://docs.transifex.com/setup/translation-memory/enabling-autofill#fill-up-resources%C2%A0from-tm-on-demand. It should help us too, especially now.

**Comment by Reinmar** [comment]
![image](https://user-images.githubusercontent.com/156149/37353999-e6836b98-26e0-11e8-8c6f-659a9146f92b.png)

**Comment by Reinmar** [comment]
And there's some information here that may help us: https://docs.transifex.com/projects/updating-content/#using-the-api-or-command-line-client-to-update-source-files

**Comment by Reinmar** [comment]
I'm thinking right now that:

* We need to understand what Translation Memory autofill is capable off. We need to know that to decide when the upload script should fail. E.g. if the autofill feature handles situations when msgCtx has changed, but msgId stayed the same, then we can allow uploading PO files which replace translations with new translations which meet that criteria. But we should block uploading PO files which e.g. replace a translation with a new one which e.g. has a bit different msgId. 
* However, to implement the safety checks like described above, the Tx's library needs to support doing a "dry-run" or something else which will allow us to understand what will happen if we'll send the new PO files for real.
* All cases blocked by the upload script need to be then handled manually. That's the only thing we can do to avoid losing existing translations if we've done some refactoring in the code or changed msgIds (to improve English wording), etc. In such case, I think that the process would look like this:
   * source PO file upload was blocked because a dangerous situation was detected,
   * so, you make changes manually in local PO files (source + translations),
   * you execute the script for manually sending all PO files (https://github.com/ckeditor/ckeditor5-dev/pull/378)
   * you do the rest of the cleanup manually – e.g. remove unused strings from Tx manually (or, the source PO file upload tool could ask if you want to do that; in step 1. you had chosen to abort, but now you can choose to proceed)

tl;dr The source PO file upload tool should ask for permission if it detects that any string will be deleted on Tx.

**Comment by ma2ciek** [comment]
We can download source files to the memory first and compare all strings and contexts with the current to be sure that everything is ok before uploading source files.

**Comment by ma2ciek** [comment]
> We need to understand what Translation Memory autofill is capable off. We need to know that to decide when the upload script should fail. E.g. if the autofill feature handles situations when msgCtx has changed, but msgId stayed the same, then we can allow uploading PO files which replace translations with new translations which meet that criteria. But we should block uploading PO files which e.g. replace a translation with a new one which e.g. has a bit different msgId.

There were no suggestions, so I'm afraid, that there'll be no autofill option :/ But I'll check it to be sure.

**Comment by Reinmar** [comment]
> There were no suggestions, so I'm afraid, that there'll be no autofill option :/ But I'll check it to be sure.

I think that's because we have that option disabled. So there was no Translation Memory, so there were no suggestions.

**Comment by Reinmar** [comment]
> And I've found that one: https://docs.transifex.com/setup/translation-memory/enabling-autofill#fill-up-resources%C2%A0from-tm-on-demand. It should help us too, especially now.

Uh... success story :)

I upload translations today. Then, I realised that `contexts.json` for `ckeditor5-alignment` has been changed some time ago, so when uploading translations I purged the old ones.

So, I changed the mentioned setting and upload translations again. Transifex recovered those old strings which we lost for a moment :)

It seems that we don't have to be that worried about changing `contexts.json`.

**Comment by Reinmar** [comment]
#378 contains @ma2ciek PR (closed without merging: see https://github.com/ckeditor/ckeditor5-dev/pull/378#issuecomment-528234500) which introduced tools to deal with those situations.

Code: https://github.com/ckeditor/ckeditor5-dev/compare/t/376.
